launcher/src/main/java/org/apache/spark/launcher/SparkLauncher.java

```diff
@@ -0,0 +1,371 @@
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Launcher for Spark applications.
+ * <p/>
+ * Use this class to start Spark applications programatically. The class uses a builder pattern
+ * to allow clients to configure the Spark application and launch it as a child process.
+ * <p/>
+ * There's also support for running the application on a separate thread, although that is to
+ * be considered experimental and avoided in production environments.
+ * <p/>
+ * Note that launching Spark applications using this class will not automatically load environment
+ * variables from the "spark-env.sh" or "spark-env.cmd" scripts in the configuration directory.
+ */
+public class SparkLauncher extends AbstractLauncher<SparkLauncher> {
+
+  private static final AtomicLong THREAD_ID = new AtomicLong();
+
+  protected boolean verbose;
+  protected String appName;
+  protected String master;
+  protected String deployMode;
+  protected String userClass;
+  protected String userResource;
+  protected final List<String> sparkArgs;
+  protected final List<String> userArgs;
+  protected final List<String> jars;
+  protected final List<String> files;
+  protected final List<String> pyFiles;
+
+  public SparkLauncher() {
+    this.sparkArgs = new ArrayList<String>();
+    this.userArgs = new ArrayList<String>();
+    this.jars = new ArrayList<String>();
+    this.files = new ArrayList<String>();
+    this.pyFiles = new ArrayList<String>();
+  }
+
+  /** Set the application name. */
+  public SparkLauncher setAppName(String appName) {
+    checkNotNull(appName, "appName");
+    this.appName = appName;
+    return this;
+  }
+
+  /** Set the Spark master for the application. */
+  public SparkLauncher setMaster(String master) {
+    checkNotNull(master, "master");
+    this.master = master;
+    return this;
+  }
+
+  /** Set the deploy mode for the application. */
+  public SparkLauncher setDeployMode(String mode) {
+    checkNotNull(mode, "mode");
+    this.deployMode = mode;
+    return this;
+  }
+
+  /**
+   * Set the main application resource. This should be the location of a jar file for Scala/Java
+   * applications, or a python script for PySpark applications.
+   */
+  public SparkLauncher setAppResource(String path) {
+    checkNotNull(path, "path");
+    this.userResource = path;
+    return this;
+  }
+
+  /** Sets the application class name for Java/Scala applications. */
+  public SparkLauncher setClass(String userClass) {
+    checkNotNull(userClass, "userClass");
+    this.userClass = userClass;
+    return this;
+  }
+
+  /** Adds command line arguments for the application. */
+  public SparkLauncher addArgs(String... args) {
+    for (String arg : args) {
+      checkNotNull(arg, "arg");
+      userArgs.add(arg);
+    }
+    return this;
+  }
+
+  /** Adds a jar file to be submitted with the application. */
+  public SparkLauncher addJar(String path) {
+    checkNotNull(path, "path");
```

Comment:
should the second argument be "jar" instead? Otherwise we get something like
```
'path' must not be null
```
and it's not clear whether we're adding a jar or a file